File: ACTransit.RestroomFinder.Native/Apple/Restroom/ServerAPI.swift
//
//  ServerAPI.swift
//  Restroom
//
//  Created by DevTeam on 12/18/15.
//  Copyright © 2015 DevTeam. All rights reserved.
//

import Foundation
import UIKit
import MapKit
import SharedFramework

class ServerAPI: ServerBaseAPI {
    //let url:String="http://10.1.200.244/ToolBeltAPI/api/RestStops"
        
    func getRestrooms(_ route: String, location: CLLocation, isPublic:Bool, callBack: @escaping ([RestStopModel], NSError?) -> Void){
        getRestStops(route, location: location, isPublic: isPublic, callBack: callBack)
    }
    
    private func getRestStops(_ route: String, location: CLLocation, isPublic: Bool, callBack: @escaping ([RestStopModel], NSError?) -> Void){
        let lat=location.coordinate.latitude
        let long=location.coordinate.longitude
        var r = route
        if (r == "" || r == "-1" || r == "All"){
            r = ""
        }
        var surl=getRestroomURL(r,lat:String(lat), long:String(long),distance:80000, isPublic: isPublic, isPending: false)
        
        NSLog("surl: %@", surl)
        var promises : Array<AjaxPromise<[RestStopModel]>>=[]
        let res1 = super.get(surl, headers: headers) as AjaxPromise<[RestStopModel]>
        promises.append(res1)
//        res1.always.addHandler({p in
//            if (p.data==nil){
//                callBack([],p.error)
//            }
//            else{
//                callBack(p.data, p.error)
//            }
//        })
        
        if (!isPublic && Common.canAddRestroom()){
            surl=getRestroomURL(r,lat:String(lat), long:String(long),distance:80000, isPublic: isPublic, isPending: true)
            let res2 = super.get(surl, headers: headers) as AjaxPromise<[RestStopModel]>
            promises.append(res2)
        }
        
        let pHandler = PromiseHandler(promises: promises);
        pHandler.always.addHandler({arr in
            var err:NSError!
            var data:[RestStopModel] = []
            for p in arr{
                
                err = p.error
                if (err != nil){
                    break;
                }
                if (p.data != nil){
                    data.append(contentsOf: p.data)
                }
            }
            if (err != nil){
                callBack([], err)
            }
            else{
                callBack(data, err)
            }
                
        })
        pHandler.wait()
    }
    
    public func getRestStop(_ restStopId: Int, isPublic: Bool, callBack: @escaping (RestStopModel?, NSError?) -> Void){
//        let surl=getRestroomURL(restStopId, isPublic: isPublic)
//        let res = super.get(surl, headers: headers) as AjaxPromise<[RestStopModel]>
//        res.always.addHandler({p in
//            var data: RestStopModel?
//            if (p.data != nil && p.data.count>0){
//                data = p.data[0]
//            }
//            callBack(data,p.error)
//        })
        if (Common.currentLocation == nil) {
            callBack(nil, NSError(domain: "ApplicationGenerated", code: 401,
                                  userInfo: [NSLocalizedDescriptionKey:NSLocalizedString("UnknownLocation", comment: "Location is unknown")]))
            return
        }
        self.getRestrooms("All", location: Common.currentLocation!, isPublic: false, callBack: {(jsonResult:[RestStopModel], error:NSError!) -> Void in
            DispatchQueue.main.async(execute: {
                defer{
                    if (error != nil){
                        callBack(nil,error)
                    }
                }
                for r in jsonResult {
                    if (r.id == restStopId){
                        callBack(r,nil)
                        return
                    }
                }
                
            })
        })
    }
 
    //NOTE: New method to retrieve demo-hard-coded data for the General Office and Route 1 restrooms.
//    func getRestroomsAroundGeneralOffice(_ callBack: @escaping ([RestStopModel], NSError?) -> Void) {
//        let surl = getRestURL("1", lat: "37.805413", long: "-122.268528", distance: nil)
//        let res = super.get(surl, headers: headers) as AjaxPromise<[RestStopModel]>
//        res.always.addHandler({p in
//            if (p.data==nil){
//                callBack([],p.error)
//            }
//            else{
//                callBack(p.data, p.error)
//            }
//        })
//
//    }
    func sendSecurityCode(_ firstName: String, lastName: String, badge: String, phoneNumber: String, callBack: @escaping (Bool, NSError?) -> Void){
        let surl = Constants.Server.sendSecurityCodeUrl
        let obj = AuthenticationModel()
        obj.firstName = firstName
        obj.lastName = lastName
        obj.badge = badge
        obj.phoneNumber = phoneNumber
        obj.agreed = false
        obj.validating = false
        obj.deviceSessionId = AppStorage.sessionId
        obj.incidentDateTime = DateTime.now().toString()
        let head=headers;
        let res = super.post(surl, jsonObject: obj.toJsonObject(), headers: head) as AjaxPromise<BoolModel>
        res.always.addHandler({p in
            callBack(p.data?.value ?? false,p.error)
        })
    }
    func validateSecurityCode(_ firstName: String, lastName: String, badge: String, phoneNumber: String, securityCode: String, callBack: @escaping (AuthenticationModel?, NSError?) -> Void){
        let surl = Constants.Server.validateSecurityCodeUrl
        let obj = AuthenticationModel()
        obj.firstName = firstName
        obj.lastName = lastName
        obj.badge = badge
        obj.phoneNumber = phoneNumber
        obj.confirm2FACode = securityCode
        obj.agreed = false
        obj.validating = false
        obj.deviceSessionId = AppStorage.sessionId
        obj.incidentDateTime = DateTime.now().toString()
        let head=headers;
        let res = super.post(surl, jsonObject: obj.toJsonObject(), headers: head) as AjaxPromise<AuthenticationModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    
    
    func getOperationAsync(agreed: Bool, validating: Bool, callBack: @escaping (OperationModel?, NSError?) -> Void){
        let surl = Constants.Server.operationUrl
        let obj = OperationModel()
        obj.firstName = AppStorage.firstName
        obj.lastName = AppStorage.lastName
        obj.badge = AppStorage.badge
        obj.agreed = agreed
        //obj.cardNumber="123456"
        
        obj.validating = validating
        obj.deviceSessionId = AppStorage.sessionId
        obj.incidentDateTime = DateTime.now().toString()
        let head=headers;
        let res = super.post(surl, jsonObject: obj.toJsonObject(), headers: head) as AjaxPromise<OperationModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    
    func authenticateAsync(callBack: @escaping (AuthenticationModel?, NSError?) -> Void){
        let surl = Constants.Server.authenticateUrl
        let obj = AuthenticationModel()
        obj.firstName = AppStorage.firstName
        obj.lastName = AppStorage.lastName
        obj.badge = AppStorage.badge
        obj.phoneNumber = AppStorage.phoneNumber
        obj.confirm2FACode = ""
    
        obj.deviceSessionId = AppStorage.sessionId
        obj.incidentDateTime = DateTime.now().toString()
        let head=headers;
        let res = super.post(surl, jsonObject: obj.toJsonObject(), headers: head) as AjaxPromise<AuthenticationModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    
    
    func getTripPattern(_ route: String, tripId: Int, callBack: @escaping ([Timepoint], NSError?) -> Void) {
        let surl = getTripPatternURL(route, tripId: tripId)
        let res = super.get(surl, headers: headers) as AjaxPromise<[Timepoint]>
        res.always.addHandler({p in
            if (p.data==nil){
                callBack([],p.error)
            }
            else{
                callBack(p.data, p.error)
            }
        })
        
    }
    
    
    func getRoutes(_ callBack: @escaping ([RouteModel], NSError?) -> Void) {
        let surl = getRoutesURL()
        let res = super.get(surl, headers: nil) as AjaxPromise<[RouteModel]>
        res.always.addHandler({p in
            if (p.data==nil){
                callBack([],p.error)
            }
            else{
                callBack(p.data, p.error)
            }
        })
    }
    func getVersion(_ callBack: @escaping ([VersionModel], NSError?) -> Void) {
        let surl = Constants.Server.versionUrl
        let res = super.get(surl, headers: headers) as AjaxPromise<[VersionModel]>
        res.always.addHandler({p in
            if (p.data==nil){
                callBack([],p.error)
            }
            else{
                callBack(p.data, p.error)
            }
        })
    }
    
    func getFeedbacks(_ restroomId: Int, callBack: @escaping ([FeedbackModel], NSError?) -> Void) {
        let surl = getFeedbacksURL(restroomId)
        let res = super.get(surl, headers: headers) as AjaxPromise<[FeedbackModel]>
        res.always.addHandler({p in
            if (p.data==nil){
               callBack([],p.error)
           }
           else {
               callBack(p.data,p.error)
           }
        
        })
    }

    func saveFeedback(_ feedback: FeedbackModel, callBack: @escaping (FeedbackModel?, NSError?) -> Void) {
        let surl = getFeedbacksURL(nil)
        let res = post(surl, jsonObject: feedback.toJsonObject(), headers: headers) as AjaxPromise<FeedbackModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    func saveRestroom(_ restroom: RestStopModel, isPublic:Bool, callBack: @escaping (RestStopModel?, NSError?) -> Void) {
        let surl = getRestroomURL(nil, isPublic: isPublic)
        let res = post(surl, jsonObject: restroom.toJsonObject(), headers: headers) as AjaxPromise<RestStopModel>
        res.always.addHandler({p in
            callBack(p.data,p.error)
        })
    }
    func getTripPatternURL(_ route:String, tripId:Int) -> String{
        let tripPatternURL = "\(Constants.Server.transitApiUrl)/route/\(route)/trip/\(tripId)/pattern?token=\(Constants.Server.transitApiToken)"
        print(tripPatternURL)
        return tripPatternURL
    }
    
    //NOTE: Made distance nullable
    func getRestroomURL(_ route:String, lat:String, long:String, distance:Int?,isPublic: Bool, isPending: Bool)-> String{
        let baseUrl = isPublic ? Constants.Server.restroomUrl : (isPending ? Constants.Server.actNewPendingRestroomUrl : Constants.Server.actRestroomUrl)
        let returnUrl = "\(baseUrl)?routeAlpha=\(route)&direction=&lat=\(lat)&longt=\(long)"

        if(distance != nil) {
            return "\(returnUrl)&distance=\(String(describing: distance!))";
        } else {
            return returnUrl;
        }
    }
    func getRestroomURL(_ restStopId: Int?, isPublic: Bool)-> String{
        if (restStopId != nil){
            //return  "\(isPublic ? Constants.Server.restroomUrl : Constants.Server.actRestroomUrl)/\(restStopId!)"
            return  "\(isPublic ? Constants.Server.restroombyIdUrl : Constants.Server.restroombyIdUrl)/?id=\(restStopId!)"
        }
        return "\(isPublic ? Constants.Server.saveRestroomUrl : Constants.Server.saveActRestroomUrl)/"
    }
    func getOperationURL(_ badge:String)-> String{
        return "\(Constants.Server.operationUrl)?badgeNumber=\(badge)"
    }
    
    func getRoutesURL() -> String{
        return "\(Constants.Server.transitApiUrl)/routes/?token=\(Constants.Server.transitApiToken)"
    }
    func getFeedbacksURL(_ restroomId : Int?) -> String{
        if (restroomId != nil){
            return "\(Constants.Server.feedbackUrl)/get?restroomId=\(restroomId!)"
        }
        return "\(Constants.Server.feedbackUrl)/post"
    }
  
    var  headers : [String:String]{
        get{
            return ["sessionId":AppStorage.sessionId,
                    "deviceGuid" : Common.currentDevice.identifierForVendor!.uuidString,
                    "firstName": AppStorage.firstName,
                    "lastName": AppStorage.lastName,
                    "badge": AppStorage.badge]
        }
    }
    
}

File: ACTransit.RestroomFinder.Native/Android/RestroomFinder/app/src/main/java/org/actransit/restroomfinder/Infrastructure/ServerAPI.java
package org.actransit.restroomfinder.Infrastructure;

import org.actransit.restroomfinder.MapviewActivity;
import org.actransit.restroomfinder.Model.AuthenticationModel;
import org.actransit.restroomfinder.Model.BoolModel;
import org.actransit.restroomfinder.Model.FeedbackModel;
import org.actransit.restroomfinder.Model.OperationModel;
import org.actransit.restroomfinder.Model.RestroomModel;
import org.actransit.restroomfinder.Model.RouteModel;
import org.actransit.restroomfinder.Model.VersionModel;
import org.actransit.restroomfinder.StartupActivity;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.lang.reflect.Type;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.logging.Logger;

import android.content.Context;
import android.util.Log;

import com.android.volley.VolleyError;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.tellexperience.jsonhelper.JsonMapper;
import com.tellexperience.jsonhelper.ServerBaseAPI;

/**
 * Created by atajadod on 5/16/2016.
 */
public class ServerAPI extends ServerBaseAPI{

    private Context context;
    public ServerAPI(Context context,String sessionId){
        super(context);
        this.context=context;
        //if (sessionId!=null && !sessionId.isEmpty())
          //  headers.put("sessionId", sessionId);

        super.headers= getHeaders();
    }

    public void getRestrooms(String route, Double latitude, Double longitude, ServerResult callBack){
//        final int callbacks=0;
//        final int callbacksCount=2;
        ServerMultiResult internalCallback=new ServerMultiResult<List<RestroomModel>>() {
            int callbacks=0;
            int callbacksResolved=0;
            List<RestroomModel> finalData=new ArrayList<RestroomModel>();
            Exception finalError;

            @Override
            public void setCallbacks(int callbacks) {
                this.callbacks = callbacks;
            }
            @Override
            public void Always(final List<RestroomModel> data, Exception error) {
                callbacksResolved++;
                if (error != null)
                    finalError=error;
                else if (data != null)
                    finalData.addAll(data);
                if (callbacksResolved==callbacks)
                    callBack.Always(finalData,finalError);
            }
        };

        Log.d("get restroom", "here we are now......");

        int callbacksNumber=1;

        if (Common.Loggedin){
            if (Constants.PublicViewEnabled && (route==null || route.equals("")))
                callbacksNumber++;
            if (Common.canAddRestroom)
                callbacksNumber++;
        }
        internalCallback.setCallbacks(callbacksNumber);

        if (Common.Loggedin) {
            String url = getActRestroomsURL(route==null?"":route,latitude,longitude,80000, false);
            Get(url, RestroomModel.class, internalCallback);

            if (Common.canAddRestroom){
                url = getActRestroomsURL(route==null?"":route,latitude,longitude,80000, true);  //now get the new pending restrooms.
                Get(url, RestroomModel.class, internalCallback);
            }
        }
        if (Constants.PublicViewEnabled && (route==null || route.equals(""))){
            String publicUrl = getPublicRestroomsURL(route==null?"":route,latitude,longitude,80000);
            Get(publicUrl, RestroomModel.class, internalCallback);
        }

    }
    public void getRestroom(int restroomId, ServerResult callBack){
        String url = getRestroomURL(restroomId);
        Get(url, RestroomModel.class, callBack);
    }


    public void getRoutes(ServerResult callBack){
        String url = getRoutesURL();
        Get(url, RouteModel.class, callBack);
    }

    public void getOperation(boolean agreed,boolean  validating, ServerResult callBack)  {
        String surl = Constants.Server.operationUrl;
        OperationModel obj = new OperationModel();
        AppStorage storage= AppStorage.Current(MyApplication.getAppContext());
        obj.firstName = storage.getFirstName();
        obj.lastName = storage.getLastName();
        obj.badge = storage.getBadge();
        obj.agreed = agreed;
        obj.validating = validating;
        obj.deviceSessionId = storage.getSessionId();
        String d=getCurrentDateTime();
        obj.incidentDateTime= d;//Calendar.getInstance().getTime().toString();
        JSONObject jObj= JsonMapper.toJSON(obj, OperationModel.class);
        Post(surl, jObj, OperationModel.class, callBack);
    }

    public void getFeedbacks(int restStopId, ServerResult callBack) {
        String url = getFeedbacksURL(restStopId);
        Get(url, FeedbackModel.class, callBack);
    }
    public void getVersion(ServerResult callBack){
        String url=Constants.Server.versionUrl + "/" +  android.os.Build.VERSION.SDK_INT;
        Get(url, VersionModel.class,callBack);
    }

    public void saveFeedback(FeedbackModel feedback,ServerResult callBack) {
        String url = getFeedbacksURL(0);
        JSONObject jObj=JsonMapper.toJSON(feedback, FeedbackModel.class);
        Post(url,jObj,FeedbackModel.class ,callBack);
    }


    public void sendSecurityCode(String firstName, String lastName, String badge, String phoneNumber, ServerResult<BoolModel>  callBack){
        String surl = Constants.Server.sendSecurityCodeUrl;
        AuthenticationModel obj = new AuthenticationModel(firstName,lastName, badge, phoneNumber);
        obj.deviceSessionId=getCurrentAppStorage().getSessionId();

        String d=getCurrentDateTime();
        obj.incidentDateTime= d;//Calendar.getInstance().getTime().toString();
        JSONObject jObj=JsonMapper.toJSON(obj, AuthenticationModel.class);
        //super.headers= getHeaders(); // AppStorage not yet set.
        Post(surl,jObj,BoolModel.class,callBack);

    }
    public void  validateSecurityCode(String firstName, String lastName, String badge, String phoneNumber, String securityCode, ServerResult<AuthenticationModel> callBack){
        String surl = Constants.Server.validateSecurityCodeUrl;
        AuthenticationModel obj = new AuthenticationModel(firstName,lastName, badge, phoneNumber,securityCode);
        obj.deviceSessionId=getCurrentAppStorage().getSessionId();

        String d=getCurrentDateTime();
        obj.incidentDateTime= d;//Calendar.getInstance().getTime().toString();

        JSONObject jObj=JsonMapper.toJSON(obj, AuthenticationModel.class);
        Post(surl,jObj,AuthenticationModel.class,callBack);
    }

    public void  authenticateAsync(ServerResult<AuthenticationModel> callBack){
        String surl = Constants.Server.authenticateUrl;
        AuthenticationModel obj = new AuthenticationModel(getCurrentAppStorage().getFirstName(),
                getCurrentAppStorage().getLastName(),
                getCurrentAppStorage().getBadge(),
                getCurrentAppStorage().getPhoneNumber(), "");
        obj.deviceSessionId=getCurrentAppStorage().getSessionId();
        String d=getCurrentDateTime();
        obj.incidentDateTime= d;//Calendar.getInstance().getTime().toString();

        JSONObject jObj=JsonMapper.toJSON(obj, AuthenticationModel.class);
        Post(surl,jObj,AuthenticationModel.class,callBack);
    }

    public void saveRestroom(RestroomModel restroom, ServerResult callBack) {
        String url = Constants.Server.saveActRestroomUrl;
        JSONObject jObj=JsonMapper.toJSON(restroom, RestroomModel.class);
        Post(url,jObj,RestroomModel.class ,callBack);
    }

    private String getPublicRestroomsURL(String route, Double latitude, Double longitude, Integer distance){
        //return "http://www.android.com/";
        String returnUrl = Constants.Server.publicRestroomsUrl +  "?routeAlpha=" + route + "&direction=&lat=" + latitude.toString() + "&longt=" + longitude.toString();
        if(distance != null) {
            return returnUrl + "&distance=" + distance;
        } else {
            return returnUrl;
        }
    }
    private String getActRestroomsURL(String route, Double latitude, Double longitude, Integer distance, boolean includeNewPendings){
        //return "http://www.android.com/";
        String returnUrl = "";
        returnUrl = Constants.Server.actRestroomsUrl +  "?routeAlpha=" + route + "&direction=&lat=" + latitude.toString() + "&longt=" + longitude.toString();
        if (includeNewPendings)
            returnUrl = Constants.Server.actNewPendingRestroomsUrl +  "?routeAlpha=" + route + "&direction=&lat=" + latitude.toString() + "&longt=" + longitude.toString();
        if(distance != null) {
            return returnUrl + "&distance=" + distance;
        } else {
            return returnUrl;
        }
    }

    private String getRestroomURL(int restroomId){
        return Constants.Server.restroomByIdUrl +  "?id=" + restroomId;
    }

    private String getRoutesURL() {
        return (Constants.Server.transitApiUrl) + "/routes/?token=" + (Constants.Server.transitApiToken);
    }

    private String getFeedbacksURL(int restroomId){
        if (restroomId != 0){
            return Constants.Server.feedbackUrl + "/get?restroomId=" + restroomId;
        }

        return Constants.Server.feedbackUrl + "/Post";

    }


    private AppStorage getCurrentAppStorage(){
        return AppStorage.Current(context);
    }

    private HashMap<String, String> getHeaders(){
        HashMap<String, String> result= new HashMap<String, String>();
        AppStorage storage= getCurrentAppStorage();
        result.put("sessionId",storage.getSessionId());
        result.put("deviceGuid" , MyApplication.getDeviceId());
        result.put("firstName", storage.getFirstName());
        result.put("lastName", storage.getLastName());
        result.put("badge", storage.getBadge());
        return result;
    }

    private String getCurrentDateTime(){
        Date date = new Date();
        DateFormat dateFormat = android.text.format.DateFormat.getDateFormat(MyApplication.getAppContext());
        return dateFormat.format(date);
    }

    public String parseVolleyError(VolleyError error) {
        try {

            String responseBody = "Fail to connect to server.";
            if (error!=null && error.networkResponse!=null)
                responseBody = new String(error.networkResponse.data, "utf-8");
            String message=responseBody;
            if (responseBody!=null && responseBody.startsWith("{")){
                JSONObject data = new JSONObject(responseBody);
                if (data.has("errors")){
                    JSONArray errors = data.getJSONArray("errors");
                    JSONObject jsonMessage = errors.getJSONObject(0);
                    data=jsonMessage;
                }
                if (data.has("message"))
                    message = data.getString("message");
                else if (data.has("errorMessage"))
                    message = data.getString("errorMessage");
                else if (data.has("error"))
                    message = data.getString("error");

            }
            return message;
            //Toast.makeText(getApplicationContext(), message, Toast.LENGTH_LONG).show();
        } catch (JSONException e) {
        } catch (UnsupportedEncodingException errorr) {
        }
        return "";
    }
}

abstract class ServerMultiResult<T> implements ServerBaseAPI.ServerResult<T>
{
    public abstract void setCallbacks(int callbacks);
}